cmd/registry/cmd/upload/bulk/discovery.go

```diff
@@ -155,23 +145,37 @@ func (task *uploadDiscoveryTask) createVersion(ctx context.Context) error {
 	return nil
 }
 
-func (task *uploadDiscoveryTask) createSpec(ctx context.Context) error {
-	if _, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
+func (task *uploadDiscoveryTask) createOrUpdateSpec(ctx context.Context) error {
+	// Use the spec size and hash to avoid unnecessary uploads.
```

Comment:
Are we optimizing for bandwidth or performance? Unless we have metrics that you used to make this decision, it might actually be faster to upload without making an additional round-trip to check if we need to reupload.

This approach probably does save some amount of bandwidth, but it's unclear to me why we're optimizing for that and how much of a difference it really makes in practice.

Reply:
The "it might actually be faster" observation might be true for very small specs, but consider that this is both a utility and sample code - it shows that when bandwidth is a consideration, hashes can be used to reduce API traffic.